### PR TITLE
rule_checker: prohibit tiflash from being a witness

### LINKED ISSUE #5747: prohibit tiflash from being a witness
## Development Task

ref https://github.com/tikv/tikv/issues/12876

### PR DESCRIPTION
**Pull request by ethercflow**
ref tikv/tikv#12876

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

### What problem does this PR solve?

Issue Number: Close #5747

### What is changed and how does it work?

```
rule_checker: prohibit tiflash from being a witness
```

### Check List

Tests

- Unit test

Code changes

- rule_checker.go

Side effects

- none

Related changes

- none

### Release note

```release-note
None
```

### CODE REVIEW

File: server/schedule/checker/rule_checker.go [hunk 1]
```diff
@@ -234,7 +234,7 @@ func (c *RuleChecker) replaceUnexpectRulePeer(region *core.RegionInfo, rf *place
 		// No matter whether witness placement rule is enabled or disabled, when peer's downtime
 		// exceeds the threshold(30min), add a witness and remove the down peer. Then witness is
 		// promoted to non-witness gradually to improve availability.
-		if status == "down" {
+		if status == "down" && !core.IsStoreContainLabel(c.cluster.GetStore(store).GetMeta(), core.EngineKey, core.EngineTiFlash) {
```
Comment (rleungx) [inline server/schedule/checker/rule_checker.go:237]: Do we need to move it to line 233?

File: server/schedule/placement/rule_manager.go [hunk 2]
```diff
@@ -211,6 +212,9 @@ func (m *RuleManager) adjustRule(r *Rule, groupID string) (err error) {
 		if !validateOp(c.Op) {
 			return errs.ErrRuleContent.FastGenByArgs(fmt.Sprintf("invalid op %s", c.Op))
 		}
+		if r.IsWitness && c.Key == core.EngineKey && slices.Contains(c.Values, core.EngineTiFlash) {
+			return errs.ErrRuleContent.FastGenByArgs("witness cann't combine with tiflash")
```
Comment (rleungx) [inline server/schedule/placement/rule_manager.go:216]:
```suggestion
			return errs.ErrRuleContent.FastGenByArgs("witness can't combine with tiflash")
```

File: server/schedule/checker/rule_checker.go (hunk 1 above)
Comment (ethercflow) [reply server/schedule/checker/rule_checker.go]: Agree, 233 is better.

**Review by rleungx — APPROVED** [review]

**Review by JmPotato — APPROVED** [review]

**Pull request merged**